### PR TITLE
CMS fit and finish

### PR DESCRIPTION
**Pull request by abist**
Addresses issues we went over in meeting - https://github.com/microsoft/azuredatastudio/issues/5323

### CODE REVIEW

File: extensions/schema-compare/package.json [hunk 1]
```diff
@@ -1,52 +1,57 @@
 {
-  "name": "schema-compare",
```
Comment (udeeshagautam) [inline extensions/schema-compare/package.json]: seems unintentional :)

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 2]
```diff
@@ -60,8 +60,12 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
 	}
 
 	public async $openConnectionDialog(providers: string[], initialConnectionProfile?: IConnectionProfile, connectionCompletionOptions?: azdata.IConnectionCompletionOptions): Promise<azdata.connection.Connection> {
+		let connectionType: number = 1;
```
Comment (udeeshagautam) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: can we not use ConnectionType.editor

File: extensions/schema-compare/package.json (hunk 1 above)
Comment (abist) [reply extensions/schema-compare/package.json]: Doesn't seem like there were any changes
This change was probably because of the format document command.

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts (hunk 2 above)
Comment (abist) [reply src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: No, because if it's an editor connection, it expects an `input` from the params as well

File: extensions/cms/package.json [hunk 3]
```diff
@@ -544,15 +544,6 @@
 					"dark": "resources/dark/refresh_inverse.svg"
 				}
 			},
-			{
-				"command": "cms.resource.refreshCmsServer",
```
Comment (udeeshagautam) [inline extensions/cms/package.json]: so once we remove this he original "refresh" takes care of refreshing cms automatically?

File: extensions/schema-compare/package.json (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/schema-compare/package.json]: There's this _metadata tag added. Dunno what that's about but you should revert this regardless since it's not related to your changes

---
In reply to: [285718197](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r285718197) [](ancestors = 285718197)

File: extensions/cms/package.nls.json [hunk 4]
```diff
@@ -2,14 +2,13 @@
...
 	"cms.resource.explorer.title": "Central Management Servers",
 	"cms.resource.refreshServerGroup.title": "Refresh Server Group",
-	"cms.resource.refreshCmsServer.title": "Refresh Central Management Server",
 	"cms.resource.deleteRegisteredServer.title": "Delete Registered Server",
-	"cms.resource.addRegisteredServer.title": "Add Registered Server",
+	"cms.resource.addRegisteredServer.title": "New Server Registration...",
 	"cms.resource.deleteServerGroup.title": "Delete Server Group",
-	"cms.resource.addServerGroup.title": "Add Server Group",
+	"cms.resource.addServerGroup.title": "New Server Group...",
 	"cms.resource.registerCmsServer.title": "Add Central Management Server",
```
Comment (Charles-Gagnon) [inline extensions/cms/package.nls.json]: >Add [](start = 42, length = 3)

New CMS? To keep it in line with the other changes?

Comment (abist) [reply extensions/cms/package.nls.json]: I used all the strings from SSMS, except `Register Central Management Server...` which was given to me during the PM meeting.

**Comment by Charles-Gagnon** [comment]
> 	public async addRegisteredServer(relativePath: string, ownerUri: string,

Not directly related to this PR - but why is the return type any? Can you not have a useful return type? provider.AddRegisterServer returns a boolean so that seems like a more useful thing to return

---
Refers to: extensions/cms/src/apiWrapper.ts:262 in fb618c0. [](commit_id = fb618c01771efded1a8e2414efb6f0e9e64d9ace, deletion_comment = False)

File: extensions/cms/src/cmsResource/tree/cmsResourceTreeNode.ts [hunk 5]
```diff
@@ -39,10 +39,13 @@ export class CmsResourceTreeNode extends CmsResourceTreeNodeBase {
...
 		try {
 			let nodes: TreeNode[] = [];
+			if (!this.ownerUri) {
+				this._ownerUri = await this.appContext.apiWrapper.getUriForConnection(this.connection);
+			}
 			return this.appContext.apiWrapper.createCmsServer(this.connection, this.name, this.description).then(async (result) => {
 				if (result) {
 					if (result.registeredServersList) {
-						result.registeredServersList.forEach((registeredServer) => {
+						result.registeredServersList.forEach(async (registeredServer) => {
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/tree/cmsResourceTreeNode.ts]: Why did you change this to async? It's not doing any async calls

File: extensions/cms/src/cmsResource/tree/cmsResourceTreeNode.ts [hunk 6]
```diff
@@ -56,7 +59,7 @@ export class CmsResourceTreeNode extends CmsResourceTreeNodeBase {
 					if (result.registeredServerGroups) {
 						if (result.registeredServerGroups) {
 							this._serverGroupNodes = [];
-							result.registeredServerGroups.forEach((serverGroup) => {
+							result.registeredServerGroups.forEach(async (serverGroup) => {
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/tree/cmsResourceTreeNode.ts]: Same, the async seems unnecessary.

File: extensions/cms/src/cmsResource/tree/treeProvider.ts [hunk 7]
```diff
@@ -35,18 +35,18 @@ export class CmsResourceTreeProvider implements TreeDataProvider<TreeNode>, ICms
 				// Call to collect all locally saved CMS servers
 				// to determine whether the system has been initialized.
 				let cmsConfig = this._appContext.apiWrapper.getConfiguration();
-				let cachedServers = cmsConfig ? cmsConfig.cmsServers : [];
+				let cachedServers = cmsConfig.servers ? cmsConfig.servers : [];
 				if (cachedServers && cachedServers.length > 0) {
 					let servers = [];
-					cachedServers.forEach((server) => {
+					cachedServers.forEach(async (server) => {
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/tree/treeProvider.ts]: Same, why async?

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: I think she's asking - and I agree with - is why aren't you using the enum you made in connectionManagement.ts? Hardcoded numbers like this are bleh. 

---
In reply to: [285718562](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r285718562) [](ancestors = 285718562)

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 8]
```diff
@@ -70,16 +74,16 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
...
-			connectionProfile.saveProfile = initialConnectionProfile.saveProfile;
-			await this._connectionManagementService.connectAndSaveProfile(connectionProfile, undefined, {
-				saveTheConnection: isUndefinedOrNull(connectionCompletionOptions.saveConnection) ? true : connectionCompletionOptions.saveConnection,
-				showDashboard: isUndefinedOrNull(connectionCompletionOptions.showDashboard) ? false : connectionCompletionOptions.showDashboard,
-				params: undefined,
-				showConnectionDialogOnError: isUndefinedOrNull(connectionCompletionOptions.showConnectionDialogOnError) ? true : connectionCompletionOptions.showConnectionDialogOnError,
-				showFirewallRuleOnError: isUndefinedOrNull(connectionCompletionOptions.showFirewallRuleOnError) ? true : connectionCompletionOptions.showFirewallRuleOnError
-			});
+			if (connectionCompletionOptions.saveConnection) {
+				await this._connectionManagementService.connectAndSaveProfile(connectionProfile, undefined, {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: You're changing behavior here - now if someone sets saveProfile on the initialConnectionProfile they won't actually end up saving the connection unless they also specify connectionCompletionOptions.saveConnection.

I think you're on the right track (save connection doesn't seem like it should belong in connectionProfile) but let's make sure we aren't breaking existing behavior here and ideally actually remove saveConnection from connectionProfile entirely.

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 9]
```diff
@@ -70,16 +74,16 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
 
 		if (connectionCompletionOptions) {
 			// Somehow, connectionProfile.saveProfile is false even if initialConnectionProfile.saveProfile is true, reset the flag here.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: This comment no longer applies with your changes

File: src/sql/workbench/services/connection/browser/cmsConnectionWidget.ts [hunk 10]
```diff
@@ -97,10 +97,10 @@ export class CmsConnectionWidget extends ConnectionWidget {
 		let newAuthTypes = authTypeOption.categoryValues;
 		// True when opening a CMS dialog to add a registered server
 		if (authTypeChanged) {
-			// Need to filter out SQL Login because registered servers don't support it
-			newAuthTypes = authTypeOption.categoryValues.filter((option) => option.name !== AuthenticationType.SqlLogin);
+			// Need to filter out SQL Login and AAD because registered servers don't support it
+			newAuthTypes = authTypeOption.categoryValues.filter((option) => option.name === AuthenticationType.Integrated);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/cmsConnectionWidget.ts]: I'd just say what this is actually doing - that CMS only supports Integrated Auth so filter to just that.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 11]
```diff
@@ -302,9 +303,20 @@ export class ConnectionDialogService implements IConnectionDialogService {
 
 	private handleShowUiComponent(input: OnShowUIResponse) {
 		if (input.selectedProviderType) {
-			this._currentProviderType = input.selectedProviderType;
+			// If the call is for specific providers
+			if (this._params && this._params.providers) {
+				this._params.providers.forEach((provider) => {
+					if (input.selectedProviderType === this._providerNameToDisplayNameMap[provider]) {
+						this._currentProviderType = provider;
+					}
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts]: So we're fine with this being unchanged if the provider type doesn't exist in _params.providers? What does that even mean? 

Also shouldn't you be doing something (clearing the value?) if the name doesn't exist in the params?

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts (hunk 8 above)
Comment (abist) [reply src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: Added back the `connectionProfile.saveProfile = initialConnectionProfile.saveProfile;`

I agree that we should remove `saveProfile` from `ConnectionProfile`, but that change would take a lot of API changes, which I want to avoid for now until the Connection refactor.

**Comment by kburtram** [comment]
@abist it would be great to see some test updates with this PR.  It looks like there are numerous functional updates across several feature areas and it's a bit difficult to parse if they're all correct.  Some tests would help.

**Comment by kburtram** [comment]
@abist what's the plan with this PR?  We should try to get a June release candidate extension ready soon for sign-off testing.

**Comment by abist** [comment]
@kburtram I'm trying to write integration tests but I'm hitting issues in running any integration tests at all. I synced with @alanrenmsft  yesterday, and couldn't figure out the issue. Will be working on this today.

File: extensions/cms/package.json [hunk 12]
```diff
@@ -518,7 +518,7 @@
 			{
 				"command": "cms.resource.deleteRegisteredServer",
 				"title": "%cms.resource.deleteRegisteredServer.title%",
-				"when": "viewItem == cms.resource.itemType.registeredServer"
+				"when": "viewItem == cms.resource.itemType.databaseServer"
```
Comment (kburtram) [inline extensions/cms/package.json]: where the name `databaseServer` come from?  this is a bit confusing to me since we already work with "databases" and "servers".

Comment (abist) [reply extensions/cms/package.json]: This is because right now the way a node looks for context actions is by looking at "database" and "servers" in a regex in the itemType.

File: extensions/cms/src/apiWrapper.ts [hunk 13]
```diff
@@ -106,7 +108,7 @@ export class ApiWrapper {
 	}
 
 	public async setConfiguration(value: any): Promise<void> {
-		await vscode.workspace.getConfiguration('cms').update('cmsServers', value, true);
+		await vscode.workspace.getConfiguration('cms').update('servers', value, true);
```
Comment (kburtram) [inline extensions/cms/src/apiWrapper.ts]: I wonder if the settings should be under a node `centralManagementServers`?  Do we use acronyms for other settings?

File: extensions/cms/src/apiWrapper.ts [hunk 14]
```diff
@@ -184,6 +185,17 @@ export class ApiWrapper {
 		return azdata.connection.openConnectionDialog(providers, initialConnectionProfile, connectionCompletionOptions);
 	}
 
+	public async getUriForConnection(connection: azdata.connection.Connection): Promise<string> {
```
[the remark this answers is not in the document]
Comment (kburtram) [reply extensions/cms/src/apiWrapper.ts]: any thoughts on this comment?  it doesn't appear to have been addressed.

File: extensions/cms/src/apiWrapper.ts [hunk 15]
```diff
@@ -269,20 +282,17 @@ export class ApiWrapper {
 				authTypeChanged: true
 			}
 		};
-		return this.openConnectionDialog(['MSSQL-CMS'], initialProfile, undefined).then((connection) => {
+		return this.openConnectionDialog([cmsProvider], initialProfile, { saveConnection: false }).then(async (connection) => {
```
Comment (kburtram) [inline extensions/cms/src/apiWrapper.ts]: what does `saveConnection: false` do here?  It seems like we are saving the connection the CMS view, so the parameter seems a little unintuitive.

Comment (abist) [reply extensions/cms/src/apiWrapper.ts]: The CMS server always gets saved locally in a different setting. This is for not saving into the OE and `Saved Connections` since if it explicitly set to false, they will be saved.

File: extensions/cms/src/apiWrapper.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (abist) [reply extensions/cms/src/apiWrapper.ts]: Updating with new utils to trigger a build

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 16]
```diff
@@ -60,26 +60,31 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
 	}
 
 	public async $openConnectionDialog(providers: string[], initialConnectionProfile?: IConnectionProfile, connectionCompletionOptions?: azdata.IConnectionCompletionOptions): Promise<azdata.connection.Connection> {
+		let connectionType = ConnectionType.editor;
```
Comment (kburtram) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: I'm not really sure what you're doing here?  the editor shouldn't be connecting the connection dialog through the extensibility APIs, should it?

File: src/sql/workbench/api/node/sqlExtHost.api.impl.ts [hunk 17]
```diff
@@ -121,8 +121,8 @@ export function createApiFactory(
 				getUriForConnection(connectionId: string): Thenable<string> {
 					return extHostConnectionManagement.$getUriForConnection(connectionId);
 				},
-				connect(connectionProfile: azdata.IConnectionProfile): Thenable<azdata.ConnectionResult> {
-					return extHostConnectionManagement.$connect(connectionProfile);
+				connect(connectionProfile: azdata.IConnectionProfile, saveConnection: boolean, showDashboard: boolean): Thenable<azdata.ConnectionResult> {
```
Comment (kburtram) [inline src/sql/workbench/api/node/sqlExtHost.api.impl.ts]: are these changes back-compat?  e.g., does postgres still work, etc.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 18]
```diff
@@ -235,14 +235,15 @@ export class ConnectionDialogService implements IConnectionDialogService {
...
+		let fromExtension = params && params.connectionType === ConnectionType.extension;
 		let uri: string = undefined;
 		if (fromEditor && params && params.input) {
 			uri = params.input.uri;
 		}
 		let options: IConnectionCompletionOptions = this._options || {
 			params: params,
-			saveTheConnection: true,
-			showDashboard: params && params.showDashboard !== undefined ? params.showDashboard : !fromEditor,
+			saveTheConnection: !fromExtension,
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts]: I'm not sure why "fromExtension" means don't save?

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts (hunk 16 above)
Comment (kburtram) [reply src/sql/workbench/api/node/mainThreadConnectionManagement.ts]:
```TypeScript
let connectionType = ConnectionType.editor;
if (connectionCompletionOptions && !connectionCompletionOptions.saveConnection) {
	connectionType = ConnectionType.extension;
}
```

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 19]
```diff
@@ -60,26 +60,31 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
...
-			// Somehow, connectionProfile.saveProfile is false even if initialConnectionProfile.saveProfile is true, reset the flag here.
-			connectionProfile.saveProfile = initialConnectionProfile.saveProfile;
-			await this._connectionManagementService.connectAndSaveProfile(connectionProfile, undefined, {
-				saveTheConnection: isUndefinedOrNull(connectionCompletionOptions.saveConnection) ? true : connectionCompletionOptions.saveConnection,
-				showDashboard: isUndefinedOrNull(connectionCompletionOptions.showDashboard) ? false : connectionCompletionOptions.showDashboard,
-				params: undefined,
-				showConnectionDialogOnError: isUndefinedOrNull(connectionCompletionOptions.showConnectionDialogOnError) ? true : connectionCompletionOptions.showConnectionDialogOnError,
-				showFirewallRuleOnError: isUndefinedOrNull(connectionCompletionOptions.showFirewallRuleOnError) ? true : connectionCompletionOptions.showFirewallRuleOnError
-			});
+			if (connectionCompletionOptions.saveConnection) {
```
Comment (kburtram) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]:
```TypeScript
// Somehow, connectionProfile.saveProfile is false even if initialConnectionProfile.saveProfile is true, reset the flag here.	
if (connectionCompletionOptions && connectionCompletionOptions.saveConnection) {
```

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 20]
```diff
@@ -404,7 +419,8 @@ export class ConnectionDialogService implements IConnectionDialogService {
 		return new Promise<void>((resolve, reject) => {
 			this.updateModelServerCapabilities(model);
 			// If connecting from a query editor set "save connection" to false
-			if (params && params.input && params.connectionType === ConnectionType.editor) {
+			if (params && params.input && params.connectionType === ConnectionType.editor ||
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts]: Change to this to avoid potential nullref if params in undefined. 
```TypeScript
if (params && params.input && (params.connectionType === ConnectionType.editor ||
				params.connectionType === ConnectionType.extension)) {
```

File: extensions/cms/package.json (hunk 12 above)
Comment (Charles-Gagnon) [reply extensions/cms/package.json]: Could you explain this some more? Why was registeredServer causing a problem?

---
In reply to: [289212851](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r289212851) [](ancestors = 289212851)

File: extensions/cms/src/apiWrapper.ts (hunk 13 above)
Comment (Charles-Gagnon) [reply extensions/cms/src/apiWrapper.ts]: Yeah - the section name definitely should be updated. Right now it's Cms: Cms Servers which isn't great at all

---
In reply to: [289213200](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r289213200) [](ancestors = 289213200)

File: extensions/cms/src/apiWrapper.ts [hunk 21]
```diff
@@ -156,7 +158,7 @@ export class ApiWrapper {
 	}
 
 	public showWarningMessage(message: string, ...items: string[]): Thenable<string | undefined> {
-		return vscode.window.showWarningMessage(message, ...items);
+		return vscode.window.showWarningMessage(message, { modal: true }, ...items);
 	}
```
Comment (Charles-Gagnon) [inline extensions/cms/src/apiWrapper.ts]: Why is this one the only modal one?

File: extensions/cms/src/apiWrapper.ts (hunk 15 above)
Comment (Charles-Gagnon) [inline extensions/cms/src/apiWrapper.ts]: >saveConnection [](start = 68, length = 14)

As discussed please add comments for each of these kind of things so we can track why things are being set in a certain way (especially with a confusing name like saveConnection)

File: extensions/cms/package.json [hunk 22]
```diff
@@ -24,8 +24,8 @@
 			"type": "object",
 			"title": "%cms.title%",
 			"properties": {
-				"cms.cmsServers": {
-					"type": "object"
+				"cms.servers": {
+					"type": "array"
```
Comment (Charles-Gagnon) [inline extensions/cms/package.json]: Suggest changing this to centralManagementServers.servers

File: extensions/cms/src/cmsResource/cmsResourceCommands.ts [hunk 23]
```diff
@@ -89,11 +89,18 @@ export function deleteRegisteredServerCommand(appContext: AppContext, tree: CmsR
 		if (!(node instanceof RegisteredServerTreeNode)) {
 			return;
 		}
-		appContext.apiWrapper.removeRegisteredServer(node.name, node.relativePath, node.ownerUri).then((result) => {
-			if (result) {
-				tree.notifyNodeChanged(undefined);
-			}
-		});
+		appContext.apiWrapper.showWarningMessage(
+			`${localize('cms.confirmDelete', 'Are you sure you want to delete ')}${node.name}?`,
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/cmsResourceCommands.ts]: >  [](start = 68, length = 1)

Suggest leaving the space outside the localized bit - it'd be easier to mistakenly leave the space out when localizing these strings

File: extensions/cms/src/apiWrapper.ts (hunk 21 above)
Comment (Charles-Gagnon) [reply extensions/cms/src/apiWrapper.ts]: (while I understand why you want this to be modal I'm more suggesting you add this as a param to the method. Sure it may only ever be passed as true currently but at least it shows intent and if future devs add calls to this they won't be surprised when the dialog is mysteriously modal)

---
In reply to: [289261130](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r289261130) [](ancestors = 289261130)

**Comment by Charles-Gagnon** [comment]
> 			light: this.appContext.extensionContext.asAbsolutePath('resources/light/folder.svg')

I just looked around and it looks like we have a folder_inverse.svg that should be used for dark theme

---
Refers to: extensions/cms/src/cmsResource/tree/serverGroupTreeNode.ts:88 in 97d89b4. [](commit_id = 97d89b493c73218009d0eb509d25dd5fd9773515, deletion_comment = False)

File: extensions/cms/src/cmsResource/tree/treeProvider.ts [hunk 24]
```diff
@@ -63,7 +63,12 @@ export class CmsResourceTreeProvider implements TreeDataProvider<TreeNode>, ICms
 			if (registeredCmsServers && registeredCmsServers.length > 0) {
 				this.isSystemInitialized = true;
 				// save the CMS Servers for future use
-				await this._appContext.apiWrapper.setConfiguration(registeredCmsServers);
+				let toSaveCmsServers = JSON.parse(JSON.stringify(registeredCmsServers));
+				toSaveCmsServers.forEach((server) => {
+					server.ownerUri = undefined,
+						server.connection.options.password = '';
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/cms/src/cmsResource/tree/treeProvider.ts]: Also please add a comment explaining that we're stripping the password out on purpose since it's saved to disk

---
In reply to: [285771274](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r285771274) [](ancestors = 285771274)

**Comment by Charles-Gagnon** [comment]
> 		return azdata.connection.getActiveConnections();

nit: How many of these are actually used? It would be great to clean this up to remove any unused functions

---
Refers to: extensions/cms/src/apiWrapper.ts:56 in 97d89b4. [](commit_id = 97d89b493c73218009d0eb509d25dd5fd9773515, deletion_comment = False)

File: extensions/cms/src/cmsResource/tree/treeProvider.ts (hunk 7 above)
Comment (Charles-Gagnon) [reply extensions/cms/src/cmsResource/tree/treeProvider.ts]: Still want to know if this is necessary, it doesn't seem to be

---
In reply to: [285770494](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r285770494) [](ancestors = 285770494)

**Comment by Charles-Gagnon** [comment]
> 		return vscode.workspace.getConfiguration('cms');

nit: Update the comment. I'd also suggest to 

1. Rename this one to getCmsServersFromConfig and then have it actually return the servers
2. Rename the setConfiguration to saveCmsServersToConfig(server: ICmsResourceNodeInfo[])

If you had more settings it might be useful to keep it more generic but as it is I think being very clear with what this is doing will be more beneficial

---
Refers to: extensions/cms/src/apiWrapper.ts:107 in 97d89b4. [](commit_id = 97d89b493c73218009d0eb509d25dd5fd9773515, deletion_comment = False)

**Comment by Charles-Gagnon** [comment]
> 	private static readonly loadingLabel = localize('cms.resource.tree.treeProvider.loadingLabel', 'Loading ...');

nit: Please move all member vars like this to the top - that's the general style we use across the app

---
Refers to: extensions/cms/src/cmsResource/tree/treeProvider.ts:107 in 97d89b4. [](commit_id = 97d89b493c73218009d0eb509d25dd5fd9773515, deletion_comment = False)

File: extensions/cms/src/cmsResource/tree/treeProvider.ts [hunk 25]
```diff
@@ -35,18 +35,18 @@ export class CmsResourceTreeProvider implements TreeDataProvider<TreeNode>, ICms
 				// Call to collect all locally saved CMS servers
 				// to determine whether the system has been initialized.
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/tree/treeProvider.ts]: So why is it necessary to do this check here inside getChildren? Shouldn't you load the servers from the configuration once when the extension is loaded and then getChildren can just use the in-memory cache?

File: extensions/cms/src/cmsResource/tree/treeProvider.ts [hunk 26]
```diff
@@ -63,18 +63,22 @@ export class CmsResourceTreeProvider implements TreeDataProvider<TreeNode>, ICms
...
+				await this._appContext.apiWrapper.setConfiguration(toSaveCmsServers);
 				return registeredCmsServers.map((server) => {
 					return new CmsResourceTreeNode(
 						server.name,
 						server.description,
 						server.ownerUri,
 						server.connection,
 						this._appContext, this, null);
-				});
+				}).sort();
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsResource/tree/treeProvider.ts]: >.sort(); [](start = 6, length = 8)

Does this sort do the right thing with regards to using the names to do the sorting?

**Comment by Charles-Gagnon** [comment]
> 	let queryProvider = azdata.dataprotocol.getProvider<azdata.QueryProvider>('MSSQL', azdata.DataProviderType.QueryProvider);

nit: Please clean up unused functions like this (although thinking about it now, this general cleanup stuff I'd prefer you do in a separate PR actually since this one is already big enough)

---
Refers to: extensions/integration-tests/src/utils.ts:51 in 97d89b4. [](commit_id = 97d89b493c73218009d0eb509d25dd5fd9773515, deletion_comment = False)

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts [hunk 27]
```diff
@@ -60,26 +60,31 @@ export class MainThreadConnectionManagement implements MainThreadConnectionManag
 	}
 
 	public async $openConnectionDialog(providers: string[], initialConnectionProfile?: IConnectionProfile, connectionCompletionOptions?: azdata.IConnectionCompletionOptions): Promise<azdata.connection.Connection> {
+		let connectionType = ConnectionType.editor;
+		if (connectionCompletionOptions && connectionCompletionOptions.saveConnection !== undefined) {
+			connectionType = connectionCompletionOptions.saveConnection ? ConnectionType.editor : ConnectionType.extension;
+		}
```
Comment (Charles-Gagnon) [inline src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: I think one big thing I'd like to fix before merging this in is figuring out a better name to give this or to move it into a separate property. It looks like we currently use this primarily (only?) in connectionDialogService.ts and while 2 of the uses are just for whether we save the connection there's an annoying one that uses it for handling connection cancel calls. It'd be good to dig into this a little more and see if we can clean it up before merging this in.

File: src/sql/workbench/api/node/sqlExtHost.api.impl.ts (hunk 17 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/api/node/sqlExtHost.api.impl.ts]: It'd be good to make these be optional

---
In reply to: [289215111](https://github.com/microsoft/azuredatastudio/pull/5542#discussion_r289215111) [](ancestors = 289215111)

File: src/sql/workbench/services/connection/browser/connectionDialogWidget.ts [hunk 28]
```diff
@@ -119,12 +119,17 @@ export class ConnectionDialogWidget extends Modal {
...
 			}
 		} else {
-			filteredProviderTypes = filteredProviderTypes.filter(x => x !== Constants.cmsProviderDisplayName);
+			filteredProviderTypes = filteredProviderTypes.filter(x => x !== Constants.cmsProviderName);
 		}
-		this._providerTypeSelectBox.setOptions(filteredProviderTypes);
+		this._providerTypeSelectBox.setOptions(filteredProviderTypes.filter((providerType, index) =>
+			// Remove duplicate listings
+			filteredProviderTypes.indexOf(providerType) === index)
+		);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionDialogWidget.ts]: Why would there be duplicate providerTypes listed here anyways?

Comment (abist) [reply src/sql/workbench/services/connection/browser/connectionDialogWidget.ts]: Because CMS and MSSQL have the same provider names

File: src/sql/workbench/api/node/sqlExtHost.api.impl.ts (hunk 17 above)
Comment (abist) [reply src/sql/workbench/api/node/sqlExtHost.api.impl.ts]: I'm fixing a bug where the API wasn't updated properly. I'm not adding this.

File: extensions/cms/src/cmsResource/tree/treeProvider.ts (hunk 25 above)
Comment (abist) [reply extensions/cms/src/cmsResource/tree/treeProvider.ts]: No, we wanted the updated CMS server, in case someone changed it in SSMS for example.

File: src/sql/workbench/api/node/mainThreadConnectionManagement.ts (hunk 27 above)
Comment (abist) [reply src/sql/workbench/api/node/mainThreadConnectionManagement.ts]: I've changed the name from `extension` to `temporary` to show that the connection is not going to be saved.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts (hunk 18 above)
Comment (abist) [reply src/sql/workbench/services/connection/browser/connectionDialogService.ts]: changed the ConnectionType name to `temporary` to reflect the same.

**Comment by abist** [comment]
@kburtram @Charles-Gagnon I've addressed all issues. Please have a look one last time before I merge.

File: extensions/cms/src/cmsUtils.ts [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+		return vscode.window.showErrorMessage(message, ...items);
+	}
+
+	/**
+	 * Get the configuration for a extensionName
+	 * @param extensionName The string name of the extension to get the configuration for
+	 * @param resource The optional URI, as a URI object or a string, to use to get resource-scoped configurations
+	 */
+	public getConfiguration(): vscode.WorkspaceConfiguration {
+		return vscode.workspace.getConfiguration('centralManagementServers');
```
Comment (Charles-Gagnon) [inline extensions/cms/src/cmsUtils.ts]: Wait - why is this duplicated here? It's already in the API wrapper

Comment (abist) [reply extensions/cms/src/cmsUtils.ts]: Some functions are needed to test extensions.
